src/broker/broker.c

```diff
@@ -831,60 +842,51 @@ static int create_rundir (attr_t *attrs)
         log_err ("error setting rundir broker attribute flags");
         goto done;
     }
-    if (do_cleanup)
-        cleanup_push_string (cleanup_directory_recursive, run_dir);
     rc = 0;
 done:
-    free (dir);
-    free (uri);
+    if (do_cleanup && run_dir != NULL)
+        cleanup_push_string (cleanup_directory_recursive, run_dir);
     return rc;
 }
 
-static int create_broker_rundir (struct overlay *ov, void *arg)
+static int init_local_uri_attr (struct overlay *ov, attr_t *attrs)
 {
-    attr_t *attrs = arg;
-    uint32_t rank;
-    const char *rundir;
-    const char *local_uri;
-    char *broker_rundir = NULL;
-    char *uri = NULL;
-    int rv = -1;
-
-    if (attr_get (attrs, "rundir", &rundir, NULL) < 0) {
-        log_msg ("create_broker_rundir: rundir attribute not set");
-        goto cleanup;
-    }
+    const char *uri;
 
-    rank = overlay_get_rank (ov);
-    if (asprintf (&broker_rundir, "%s/%u", rundir, rank) < 0) {
-        log_err ("create_broker_rundir: asprintf");
-        goto cleanup;
-    }
-    if (mkdir (broker_rundir, 0700) < 0) {
-        log_err ("create_broker_rundir: mkdir (%s)", broker_rundir);
-        goto cleanup;
-    }
-    if (attr_add (attrs, "broker.rundir", broker_rundir,
-                  FLUX_ATTRFLAG_IMMUTABLE) < 0) {
-        log_err ("create_broker_rundir: attr_add broker.rundir");
-        goto cleanup;
+    if (attr_get (attrs, "local-uri", &uri, NULL) < 0) {
+        uint32_t rank = overlay_get_rank (ov);
+        const char *rundir;
+        char buf[1024];
+
+        if (attr_get (attrs, "rundir", &rundir, NULL) < 0) {
+            log_msg ("rundir attribute is not set");
+            return -1;
+        }
+        if (snprintf (buf, sizeof (buf), "local://%s/local-%d",
+                      rundir, rank) >= sizeof (buf)) {
+            log_msg ("buffer overflow while building local-uri");
+            return -1;
+        }
+        if (attr_add (attrs, "local-uri", buf, FLUX_ATTRFLAG_IMMUTABLE) < 0) {
+            log_err ("setattr local-uri");
+            return -1;
+        }
     }
+    else {
+        char path[1024];
 
-    if (attr_get (attrs, "local-uri", &local_uri, NULL) < 0) {
-        if (asprintf (&uri, "local://%s/local", broker_rundir) < 0) {
-            log_err ("create_broker_rundir: asprintf (uri)");
-            goto cleanup;
+        if (strncmp (uri, "local://", 8) != 0) {
+            log_msg ("local-uri is malformed");
+            return -1;
         }
-        if (attr_add (attrs, "local-uri", uri, FLUX_ATTRFLAG_IMMUTABLE) < 0) {
-            log_err ("create_broker_rundir: attr_add (local-uri)");
-            goto cleanup;
+        if (snprintf (path, sizeof (path), "%s", uri + 8) >= sizeof (path)) {
```

Comment:
Since we're writing to `uri + 8`, should the bounds check be for `sizeof (path) - 8`?

Reply:
Reading from `uri + 8` so we check for overflow when writing to path.

Reply:
Oh duh! Oops, sorry for the think-o.